File: bench/memcompare/klio/collections.kt
// Object-graph / GC workload: build 1M records, group by key into a map of
// lists, aggregate per group, sort groups by total. Stresses the heap with many
// small live objects and a churny build phase.

class Rng(seed: Long) {
    private var s = seed and 0x7fffffffL
    fun next(): Long {
        s = (s * 1103515245L + 12345L) and 0x7fffffffL
        return s
    }
}

fun main() {
    val n = 1_000_000
    val groups = 1000
    val rng = Rng(12345L)
    val byKey = HashMap<Int, MutableList<Long>>()
    var totalSum = 0L
    for (i in 0 until n) {
        val v = rng.next()
        val k = i % groups
        byKey.getOrPut(k) { ArrayList() }.add(v)
        totalSum += v
    }
    val sums = ArrayList<Pair<Int, Long>>(byKey.size)
    for ((k, vs) in byKey) {
        var s = 0L
        for (v in vs) s += v
        sums.add(Pair(k, s))
    }
    sums.sortByDescending { it.second }
    val top = sums[0]
    println("collections: records=$n groups=${byKey.size} totalSum=$totalSum topKey=${top.first} topSum=${top.second}")
}

File: bench/memcompare/node/collections.js
// Object-graph / GC workload — see klio/collections.kt for the shared spec.
class Rng {
  constructor(seed) { this.s = seed & 0x7fffffff; }
  next() { this.s = (this.s * 1103515245 + 12345) & 0x7fffffff; return this.s; }
}
function main() {
  const n = 1000000, groups = 1000;
  const rng = new Rng(12345);
  const byKey = new Map();
  let totalSum = 0;
  for (let i = 0; i < n; i++) {
    const v = rng.next();
    const k = i % groups;
    let arr = byKey.get(k);
    if (arr === undefined) { arr = []; byKey.set(k, arr); }
    arr.push(v);
    totalSum += v;
  }
  const sums = [];
  for (const [k, vs] of byKey) {
    let s = 0;
    for (const v of vs) s += v;
    sums.push([k, s]);
  }
  sums.sort((a, b) => b[1] - a[1]);
  const top = sums[0];
  console.log(`collections: records=${n} groups=${byKey.size} totalSum=${totalSum} topKey=${top[0]} topSum=${top[1]}`);
}
main();
